### PR TITLE
Add hunting API.

### PR DESCRIPTION
**Pull request by BenAlfasi**

### CODE REVIEW

File: api-reference/beta/api/security-runhuntingquery.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+|Content-Type|application/json. Required.|
+
+## Request body
+In the request body, supply JSON representation of the parameters.
+
+The following table shows the parameters that can be used with this action.
+
+|Parameter|Type|Description|
+|:---|:---|:---|
+|query|String|The hunting query in KQL language.|
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-runhuntingquery.md:50]: Line 50:
@BenAlfasi , is this topic the appropriate reference for the KQL syntax that your API uses: https://learn.microsoft.com/en-us/azure/data-explorer/kql-quick-reference
Or this topic in the M365 Defender hunting API conceptual docs:
https://learn.microsoft.com/en-us/microsoft-365/security/defender/advanced-hunting-query-language?view=o365-worldwide
If not, please find the docs for the appropriate syntax. Then please add a second sentence to the description in line 50, with a link to the appropriate KQL docs. A description like:
```
"The hunting query in Kusto Query Language (KQL). For more information on KQL syntax, see [KQL quick reference](/azure/data-explorer/kql-quick-reference).
```

File: api-reference/beta/api/security-runhuntingquery.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ms.prod: "security"
+doc_type: apiPageType
+---
+
+# security: Run Hunting query
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Microsoft 365 Hunting feature is a query-based threat-hunting tool that lets you explore up to 30 days of raw data. You can proactively inspect events in your network to locate threat indicators and entities. The flexible access to data enables unconstrained hunting for both known and potential threats.
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-runhuntingquery.md:15]: Line 15 and GLOBAL:
@BenAlfasi , this API is advanced hunting which specifies a query in KQL, right? If that's the case, I'd suggest we refer to it as "advanced hunting" to let customers better correlate with how the admin docs describe it.

Can we start line 15 by saying:
"The Microsoft 365 Defender advanced hunting feature is a query-based..."

File: api-reference/beta/resources/security-huntingrowresult.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+One row result of the [hunting query results](../resources/security-huntingqueryresults.md).
+This class is a Dictionary of string, object.
+
+
+## JSON representation
+The following is a JSON representation of the resource.
+<!-- {
+  "blockType": "resource",
+  "@odata.type": "microsoft.graph.huntingRowResult"
```
Comment (angelgolfer-ms) [inline api-reference/beta/resources/security-huntingrowresult.md:24]: FYI I added  "openType": true to this JSON definition.
You can find more info here: https://dev.azure.com/msazure/One/_wiki/wikis/Microsoft%20Graph%20Partners/211782/Set-up-for-API-Doctor-validation?anchor=json-element-descriptions

File: api-reference/beta/resources/security-huntingrowresult.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+}
+-->
+``` json
+{
+    "Timestamp": "2020-08-30T06:38:35.7664356Z",
```
Comment (angelgolfer-ms) [inline api-reference/beta/resources/security-huntingrowresult.md:29]: Lines 29-31:
The JSON representation section is supposed to show the syntax of the type.
Because this is an open type without any properties, we shouldn't include these lines here. For a while, I was under the impression that this type always has these 3 properties (even though they are not defined in the "Properties" section).

FYI I'm going to remove these lines.

File: api-reference/beta/resources/security-huntingrowresult.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+---
+
+# huntingRowResult resource type
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+One row result of the [hunting query results](../resources/security-huntingqueryresults.md).
+This class is a Dictionary of string, object.
```
Comment (angelgolfer-ms) [inline api-reference/beta/resources/security-huntingrowresult.md:17]: FYI I added a paragraph to point out this is an open type, and the actual columns of a row result is specified in the project function that's part of the KQL query...

I don't know how I can clarify line 17 - Dan Kershaw lectured us before not to use the terms "class", "object" in REST API docs. Also, "Dictionary" and the comma after "string" are throwing me off. Can you rephrase what you intend this sentence to say please?

File: api-reference/beta/resources/security-huntingqueryresults.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+|results|[huntingRowResult](../resources/security-huntingrowresult.md) collection|The results of the hunting query.|
+
+## JSON representation
+The following is a JSON representation of the resource.
+<!-- {
+  "blockType": "resource",
+  "@odata.type": "microsoft.graph.huntingQueryResults"
+}
+-->
+``` json
```
Comment (angelgolfer-ms) [inline api-reference/beta/resources/security-huntingqueryresults.md:31]: Lines 32-59:
The JSON representation section should define the type of the resource.
Instead these lines are giving an example of what a resource of this type can be like.
FYI I updated this section. Hopefully this would get your PR out of API Doctor error jail.

File: api-reference/beta/api/security-runhuntingquery.md (hunk 2 above)
Comment (angelgolfer-ms) [reply api-reference/beta/api/security-runhuntingquery.md:15]: Even though the implicit expectation is customers go verify how KQL works, desirably we can provide a link. 
Is this appropriate (looks easy to understand): https://learn.microsoft.com/en-us/azure/data-explorer/kusto/query/

If it's too partial, can you substitute with a more appropriate source of info for KQL?

File: api-reference/beta/api/security-runhuntingquery.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+
+## Response
+
+If successful, this action returns a `200 OK` response code and a [huntingQueryResults](../resources/security-huntingqueryresults.md) in the response body.
+
+## Examples
+
+### Request
+<!-- {
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-runhuntingquery.md:61]: I added a description for the example request, to make it easier for those unfamiliar with KQL to catch on. Can you take a look please, and make edits where necessary?

File: changelog/Microsoft.M365.Defender.json [hunk 8]
```diff
@@ -345,7 +345,39 @@
 					"ChangeType": "Addition",
 					"Description": "Added the [List alerts](https://learn.microsoft.com/en-us/graph/api/security-list-alerts_v2?view=graph-rest-beta) operation.",
 					"Target": "alert"
-				}
+				},
```
Comment (angelgolfer-ms) [inline changelog/Microsoft.M365.Defender.json:348]: Lines 349-380,
@BenAlfasi , you should create a new changelog record  for these 4 changelist records, even if the API has gone live some time ago. We consider that customers generally don't know about the API unless the docs  are published (because that's the expectation we build for customers, and most other API producer teams try to sync their docs with the API release).
Just in case the terms "changelog records" and "changelist records" are unclear, you can please take a look at the example under #2, in this section of the CDK topic here: https://dev.azure.com/msazure/One/_wiki/wikis/Microsoft%20Graph%20Partners/211742/Changelog?anchor=changelog-generation-process-summary
In that example, there are 2 changelog records. The first changelog record has 2 changelist records. The second changelog record also has 2 changelist records as well. Each changelog record has its own CreateDateTime value. In the case of the changelog record for this PR, the CreateDateTime value will be set to the date that the PR will be published.

File: api-reference/beta/api/security-runhuntingquery.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+title: "security: runHuntingQuery"
+description: "Run Hunting query API"
+author: "BenAlfasi"
+ms.localizationpriority: medium
+ms.prod: "security"
+doc_type: apiPageType
+---
+
+# security: runHuntingQuery
+Namespace: microsoft.graph
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-runhuntingquery.md:11]: @BenAlfasi , I just realized all these  4 new files are in the security subnamespace.
Can you please update the following:
- The "Namespace:" line at the beginning of each of these 4 topics to specify microsoft.graph.security as the namespace value.
-  Rename the file for this runHuntingQuery as security-security-runhuntinquery.md, and fix links to this topic. the filename pattern for an action or function defined in a subnamespace is like this: {subnamespace name}-{binding_parameter_type}-{function_or_action_name}.md

File: api-reference/beta/api/security-runhuntingquery.md [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Queries a specified set of event, activity, or entity data supported by Microsoft 365 Defender to proactively look for cyber security threats.
+
+This advanced hunting feature in Microsoft 365 Defender lets you explore up to 30 days of raw data. You can proactively inspect events in your network to locate threat indicators and entities. The flexible access to data enables unconstrained hunting for both known and potential threats.
+
+This method specifies a Kusto Query Language (KQL) query which identifies a data table in the [advanced hunting schema](/microsoft-365/security/defender/advanced-hunting-schema-tables?view=o365-worldwide) to query, and a piped sequence of operators or functions to filter or search the data, and format or structure the query output in specific ways. 
+
+See more information about [KQL](/azure/data-explorer/kusto/query/).
+
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-runhuntingquery.md:22]: Before and after line 21:
FYI I added a couple of links to the admin docs to give more context:
https://learn.microsoft.com/en-us/microsoft-365/security/defender/advanced-hunting-query-emails-devices?view=o365-worldwide
https://learn.microsoft.com/en-us/microsoft-365/security/defender/advanced-hunting-overview?view=o365-worldwide

File: api-reference/beta/toc.yml [hunk 11]
```diff
@@ -14918,6 +14918,16 @@ items:
             href: resources/security-urlEvidence.md
           - name: User evidence
             href: resources/security-userEvidence.md
+        - name: Hunting
```
Comment (angelgolfer-ms) [inline api-reference/beta/toc.yml:14921]: Line 14921:
@BenAlfasi , can we call this "Advanced hunting" instead of "Hunting"? 
After reading through the admin topics for this feature, I think it would be better to refer to it as such, mapping more closely to the "advanced hunting" as opposed to "guided hunting" in the admin query builder.

File: api-reference/beta/api/security-runhuntingquery.md [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+description: "Run Hunting query API"
+author: "BenAlfasi"
+ms.localizationpriority: medium
+ms.prod: "security"
+doc_type: apiPageType
+---
+
+# security: runHuntingQuery
+Namespace: microsoft.graph
+
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-runhuntingquery.md:12]: Because this method has a bindingParameter of security.md, you should include  this method in the Methods table of security.
In the interest of time, FYI I just added a row for this method in security.md.
However, after you rename this file to security-security-runhuntingquery.md, please fix the filename in all the links to this file, including in security.md.

File: api-reference/beta/resources/security-huntingqueryresults.md [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+The results of running a [query for advanced hunting](../api/security-runhuntingquery.md).
+
+## Properties
+|Property|Type|Description|
+|:---|:---|:---|
+|schema|[singlePropertySchema](../resources/security-singlepropertyschema.md) collection|The schema for the response.|
+|results|[huntingRowResult](../resources/security-huntingrowresult.md) collection|The results of the hunting query.|
+
```
Comment (angelgolfer-ms) [inline api-reference/beta/resources/security-huntingqueryresults.md:23]: Lines 21-22 and GLOBAL:
Under the "Type" column, can you provide a fully qualified type value?
In general, for any table (like  Properties, Methods, Relationships tables in a resource topic, under the "type" or "return type" column, if the type is defined in a subnamespace, then please provide the fully qualified resource type. 

For example, if the return type of a method is huntingQueryResults, then under the Type column, please specify the value microsoft.graph.security.huntingQueryResults instead of huntingQueryResults.

File: api-reference/beta/toc.yml [hunk 14]
```diff
@@ -14918,6 +14918,16 @@ items:
...
+          items:
+          - name: Run hunting query
```
Comment (angelgolfer-ms) [inline api-reference/beta/toc.yml:14923]: @BenAlfasi , I have another question about the location of this set of API being under "Alerts and incidents" and the need to describe how customers can use these APIs together. Please see my question on Teams.

**Review by angelgolfer-ms — CHANGES_REQUESTED** [review]
HI @BenAlfasi , I made some direct updates to your PR, and added a few comments to the PR for your action as well. I posed a question in Teams for you too.
Can you please take a look?

File: api-reference/beta/api/security-runhuntingquery.md (hunk 9 above)
Comment (BenAlfasi) [reply api-reference/beta/api/security-runhuntingquery.md:11]: done

File: api-reference/beta/api/security-runhuntingquery.md (hunk 12 above)
Comment (BenAlfasi) [reply api-reference/beta/api/security-runhuntingquery.md:12]: done

File: api-reference/beta/api/security-runhuntingquery.md (hunk 10 above)
Comment (BenAlfasi) [reply api-reference/beta/api/security-runhuntingquery.md:24]: Ty!

File: api-reference/beta/api/security-runhuntingquery.md (hunk 1 above)
Comment (BenAlfasi) [reply api-reference/beta/api/security-runhuntingquery.md:58]: done

File: api-reference/beta/api/security-runhuntingquery.md (hunk 7 above)
Comment (BenAlfasi) [reply api-reference/beta/api/security-runhuntingquery.md:77]: looks great, thanks!

File: api-reference/beta/resources/security-huntingqueryresults.md (hunk 13 above)
Comment (BenAlfasi) [reply api-reference/beta/resources/security-huntingqueryresults.md:23]: done

File: api-reference/beta/resources/security-huntingqueryresults.md (hunk 6 above)
Comment (BenAlfasi) [reply api-reference/beta/resources/security-huntingqueryresults.md:31]: ty!

File: api-reference/beta/resources/security-huntingrowresult.md (hunk 5 above)
Comment (BenAlfasi) [reply api-reference/beta/resources/security-huntingrowresult.md:19]: rephrased, please let me know if ok :)

File: api-reference/beta/toc.yml (hunk 11 above)
Comment (BenAlfasi) [reply api-reference/beta/toc.yml:14921]: done

File: api-reference/beta/toc.yml (hunk 14 above)
Comment (BenAlfasi) [reply api-reference/beta/toc.yml:14923]: done

File: changelog/Microsoft.M365.Defender.json (hunk 8 above)
Comment (BenAlfasi) [reply changelog/Microsoft.M365.Defender.json]: done

File: api-reference/beta/api/security-alert-addcomment.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
+---
+title: "Add comment to alert"
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-alert-addcomment.md:2]: FYI I'm changing the filename to security-alert-post-comments.md. 
There's an existing filenaming pattern {subnamespace name}-{parententityname}-post-{navigationpropertyname}.md. I suppose we can repurpose that pattern even though comments is a property and not a nav property.

For more info:
filenaming without subnamespace: https://dev.azure.com/msazure/One/_wiki/wikis/Microsoft%20Graph%20Partners/211751/File-naming
filenaming taking subnamespace into account: https://dev.azure.com/msazure/One/_wiki/wikis/Microsoft%20Graph%20Partners/211759/OData-subnamespace?anchor=filenames

File: api-reference/beta/api/security-incident-addcomment.md [hunk 16]
```diff
@@ -0,0 +1,131 @@
+---
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-incident-addcomment.md:1]: FYI I'm changing the filename to security-incident-post-comments.md.
There's an existing filenaming pattern {subnamespace name}-{parententityname}-post-{navigationpropertyname}.md. I suppose we can repurpose that pattern even though comments is a property and not a nav property.

For more info:
filenaming without subnamespace: https://dev.azure.com/msazure/One/_wiki/wikis/Microsoft%20Graph%20Partners/211751/File-naming
filenaming taking subnamespace into account: https://dev.azure.com/msazure/One/_wiki/wikis/Microsoft%20Graph%20Partners/211759/OData-subnamespace?anchor=filenames

Comment (angelgolfer-ms) [reply api-reference/beta/api/security-incident-addcomment.md:1]: You also need to add this operation to the Methods table for the parent entity, incident.
In the interest of time so that we can wrap up this PR in a day, I'll add the method.

File: api-reference/beta/api/security-incident-addcomment.md [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+title: "Add comment to incident"
+description: "Adds a comment to the end of the incident comments list"
+ms.date: 09/09/2021
+author: "BenAlfasi"
+ms.localizationpriority: medium
+ms.prod: "security"
+doc_type: apiPageType
+---
+
+# Add comment to incident
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-incident-addcomment.md:11]: For the TOC title, we should use "Create comment", and for the topic title, "Create comment for incident".

There's this convention over the nuance of using "create" vs "add". Please see this CDK section: [When to use "Create" vs "Add" and "Delete" vs "Remove"](https://dev.azure.com/msazure/One/_wiki/wikis/Microsoft%20Graph%20Partners/211766/ToC-and-topic-title?anchor=when-to-use-%22create%22-vs-%22add%22-and-%22delete%22-vs-%22remove%22)

Again, part of this having some consistent topic title naming is so to let Graph Studio generate topic titles.

In the interest of time again, I'll make these changes. In case I misunderstood, please let me know.

File: api-reference/beta/api/security-alert-addcomment.md (hunk 15 above)
Comment (angelgolfer-ms) [reply api-reference/beta/api/security-alert-addcomment.md:2]: For the TOC title, we should use "Create comment", and for the topic title, "Create comment for alert".

There's this convention over the nuance of using "create" vs "add". Please see this CDK section: [When to use "Create" vs "Add" and "Delete" vs "Remove"](https://dev.azure.com/msazure/One/_wiki/wikis/Microsoft%20Graph%20Partners/211766/ToC-and-topic-title?anchor=when-to-use-%22create%22-vs-%22add%22-and-%22delete%22-vs-%22remove%22)

Again, part of this having some consistent topic title naming is so to let Graph Studio generate topic titles.

In the interest of time again, I'll make these changes. In case I misunderstood, please let me know.

File: api-reference/beta/api/security-incident-addcomment.md [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Adds a comment to an existing [incident](../resources/security-incident.md) based on the specified alert **id** property.
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-incident-addcomment.md:16]: Line 16:
I think you mean incident id property, not alert id property, is that right?
I am going ahead to make that update. If I got it wrong, please correct it here and in the Methods table of security-incident.md.

File: api-reference/beta/api/security-alert-addcomment.md [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+description: "Adds a comment to the end of the alert comments list"
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-alert-addcomment.md:3]: You also need to add this operation to the Methods table for the parent entity in security-alert.md.
In the interest of time so that we can wrap up this PR in a day, I'll add the method.

File: api-reference/beta/api/security-alert-addcomment.md [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+|Name|Description|
+|:---|:---|
+|Authorization|Bearer {token}. Required.|
+|Content-Type|application/json. Required.|
+
+## Request body
+
+In the request body of this API, along with the comment, provide the type of the [alertComment](../resources/security-alertcomment.md). See example below.
+
+|Property|Type|Description|
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-alert-addcomment.md:47]: Is "comment" technically a parameter in the request body?

FYI in the interest of time, I'm going to apply the structure for the "Request body" section and call it a parameter. If you have any input for me, let me know and update the topic accordingly please.

File: api-reference/beta/api/security-incident-addcomment.md [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+```
+
+## Request headers
+|Name|Description|
+|:---|:---|
+|Authorization|Bearer {token}. Required.|
+|Content-Type|application/json. Required.|
+
+## Request body
+
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-incident-addcomment.md:44]: Is "comment" technically a parameter in the request body?

FYI in the interest of time, I'm going to apply the structure for the "Request body" section and call it a parameter. If you have any input for me, let me know and update the topic accordingly please.

File: api-reference/beta/api/security-incident-addcomment.md [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+## Response
+
+If successful, this method returns a `200 OK` response code and an updated list of all [alertComment](../resources/security-alertcomment.md) of the incident.
+
+## Examples
+
+### Request
+The following is an example of a request.
+
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-incident-addcomment.md:60]: Line 61 and 78-98:
For any new example like this one, you don't need to be concerned about these lines initially. After the topic gets published, there's a tool that runs weekly and that will insert the language-dependent snippets and UI. You should actually not copy any over from other topics or else customers will see incorrect examples snippets for your new operation.
In the interest of time, I'm deleting these lines for you.

File: api-reference/beta/api/security-alert-addcomment.md [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+
+## Response
+
+If successful, this method returns a `200 OK` response code and an updated list of all [alertComment](../resources/security-alertcomment.md) of the alert.
+
+## Examples
+
+### Request
+The following is an example of a request.
+
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-alert-addcomment.md:60]: Line 61 and 78-98:
For any new example like this one, you don't need to be concerned about these lines initially. After the topic gets published, there's a tool that runs weekly and that will insert the language-dependent snippets and UI. You should actually not copy any over from other topics or else customers will see incorrect examples snippets for your new operation.
In the interest of time, I'm deleting these lines for you.

File: api-reference/beta/resources/security-api-overview.md [hunk 24]
```diff
@@ -29,6 +29,26 @@ Take immediate action to defend against threats using the [securityAction](secur
 
   > **Note:** Currently security actions only support application permissions.
 
+## Advanced hunting (preview)
+Advanced hunting is a query-based threat hunting tool that lets you explore up to 30 days of raw data. You can proactively inspect events in your network to locate threat indicators and entities. The flexible access to data enables unconstrained hunting for both known and potential threats.
```
Comment (angelgolfer-ms) [inline api-reference/beta/resources/security-api-overview.md:33]: Lines 32-50:
@BenAlfasi , I made some slight edits to this section. After I check in my commit, can you please take a look to make sure this section is still correct?

File: api-reference/beta/api/security-security-runhuntingquery.md [hunk 25]
```diff
@@ -49,15 +49,7 @@ POST /security/runHuntingQuery
...
-In the request body, supply JSON representation of the parameters.
-
-The following table shows the parameters that can be used with this action.
-
-|Parameter|Type|Description|
-|:---|:---|:---|
-|query|String|The hunting query in KQL language.|
-
-
+The hunting query in Kusto Query Language (KQL). For more information on KQL syntax, see [KQL quick reference](/azure/data-explorer/kql-quick-reference).
```
Comment (angelgolfer-ms) [inline api-reference/beta/api/security-security-runhuntingquery.md:52]: Line 52:
Hm, actually we need to specify the parameter named "query". 
Do you see a problem with that?
In the interest of time, I'll make the update, but if you see a problem, please let me know.

File: api-reference/beta/api/security-incident-addcomment.md (hunk 18 above)
Comment (BenAlfasi) [reply api-reference/beta/api/security-incident-addcomment.md:16]: correct

File: api-reference/beta/api/security-alert-addcomment.md (hunk 20 above)
Comment (BenAlfasi) [reply api-reference/beta/api/security-alert-addcomment.md:47]: yes, please see the request example

**Comment by angelgolfer-ms** [comment]
Per chat with @BenAlfasi , he reviewed my comments and requested me to go ahead and publish the PR.
